### PR TITLE
attempt to fix "get contacts": 'foreach(->ADR...'

### PR DESCRIPTION
**Pull request by eneiluj**
In reaction to the comment:

Nice new release, but there seems to be a bug in the contacts backend. If I try to load them I get the following message in my log files and the spinner goes on forever:
```
{"reqId":"XWhDd5NwMEnE5vBRUQgfxQAAAPc","level":0,"time":"2019-08-29T23:31:23+02:00","remoteAddr":"1.1.1.1","user":"user","app":"maps","method":"GET","url":"\/owncloud\/apps\/maps\/contacts","message":"Externally looked failed","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"16.0.4.1"}

{"reqId":"XWhDd5NwMEnE5vBRUQgfxQAAAPc","level":3,"time":"2019-08-29T23:34:23+02:00","remoteAddr":"1.1.1.1","user":"user","app":"PHP","method":"GET","url":"\/owncloud\/apps\/maps\/contacts","message":"Invalid argument supplied for foreach() at \/var\/www\/owncloud\/apps\/maps\/lib\/Controller\/ContactsController.php#92","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"16.0.4.1"}
```

@e-alfred Could you try this branch to check if the "foreach" error goes away?

### CODE REVIEW

**Comment by tacruc** [comment]
Looks good, lets wait if it solves the problem of @e-alfred.

**Comment by e-alfred** [comment]
Thanks for the patch, but the contacts spinner still loads foreever while showing the following error in the log file:

`{"reqId":"XWjoSFGJbJ@BXJ95n36AlwAAAQ8","level":0,"time":"2019-08-30T11:12:37+02:00","remoteAddr":"1.1.1.1","user":"user","app":"maps","method":"GET","url":"\/owncloud\/apps\/maps\/contacts","message":"Externally looked failed","userAgent":"Mozilla\/5.0 (X11; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"16.0.4.1"}
`

**Comment by eneiluj** [comment]
Could you check the browser error log?

The message you get in NC logs is just a warning which should not affect contacts loading.

**Comment by eneiluj** [comment]
@e-alfred The problem most likely comes from a javascript error so I need your browser error log accessible with the "javascript console". Just in case: in most browsers, it's accessible with the F12 key.

**Comment by e-alfred** [comment]
The error is gone now, but it still doesn't show any contacts, might be related to https://github.com/nextcloud/maps/issues/101.

**Comment by eneiluj** [comment]
@e-alfred Thanks for your feedback. I don't think it can be related with #101 because there, contacts are visible in Maps.

Potential reasons why you don't see contacts:
* the first error you got (the one fixed by this PR) messed up with the addresses database content and now Maps thinks no address is resolvable (you could try to delete all rows in `maps_address_geo` table)
* you don't have any contact with an address that Nominatim can resolved
* your server can't contact Nominatim (no internet connection, blacklisted...?)
* the bug of the year 2000 :wink:
* ...

**Comment by e-alfred** [comment]
I tried to drop all items from the maps_address_geo table and it seems they are getting filled up correctly (here is an example):

```
mysql> select * from oc_maps_address_geo;
+----+--------------------------------------------------------+--------------------------------------------------------------+-------------------------------------------------------------+------+------+-----------+
| id | object_uri                                             | adr                                                          | adr_norm                                                    | lat  | lng  | looked_up |
+----+--------------------------------------------------------+--------------------------------------------------------------+-------------------------------------------------------------+------+------+-----------+
| 26 | 13ca1129-4698-4c85-91e2-610f2ba00a2a.vcf               | 61 04 07;;Karl-Wiechert-Allee 10;Hannover;;30604;Deutschland | 610407;;karl-wiechert-allee10;hannover;;30604;deutschland   | NULL | NULL |         0 |
+----+--------------------------------------------------------+--------------------------------------------------------------+-------------------------------------------------------------+------+------+-----------+
12 rows in set (0.00 sec)

```

**Comment by eneiluj** [comment]
Ok I tried to precisely put this address in a contact and it does not get resolved neither. Then I removed the "Post office box" and it worked.

Could you try it also. In the meantime I'm checking why it does not work.

**Comment by v1r0x** [comment]
The nominatim search engine is really picky on addresses. In the old version of maps it didn't find a address `Some Street 12b` but `Some Street 12 b` (with space). This might be unrelated, but in the past I had several issues with it.

I just checked the [nominatim wiki on how to use the search](https://nominatim.org/release-docs/develop/api/Search/#parameters).
In the "structured requests" there is no parameter for post office box, so I guess it is not supported at all. The free-form query option works 
> Free-form queries are processed first left-to-right and then right-to-left if that fails.
so (I guess) it fails in both directions as it can't parse 61 04 07 into one of the supported parameters which are
- street=<housenumber> <streetname>
- city=<city>
- county=<county>
- state=<state>
- country=<country>
- postalcode=<postalcode>

Long story short: Dropping post office box seems fine to me :) But don't know how robust this is regarding the format of the initial `$adr`.

**Comment by e-alfred** [comment]
I tried to remove the post office box, but still no luck. Here is the table content without it:

`
| 27 | 13ca1129-4698-4c85-91e2-610f2ba00a2a.vcf               | ;;Karl-Wiechert-Allee 10;Hannover;;30604;Deutschland                                         | ;;karl-wiechert-allee10;hannover;;30604;deutschland                                      |       NULL |            NULL |         0 |
`

**Comment by tacruc** [comment]
@e-alfred The problem is that no external lookup is performed, or it fails (which does not mean an emptly result, but some non json result). If it would just have an empty result the last value would still be true.
Do you have any debug level logs? Is there any "Externally looked failed" or "External looked up address:" in there?

**Comment by tacruc** [comment]
@eneiluj I think we can merge this PR as it fixes some bugs.
@e-alfred Just read the fist commend again:
Cloud you visit https://nominatim.openstreetmap.org/search.php?q=;;Karl-Wiechert-Allee 10;Hannover;;30604;Deutschland&format=json
from the server you run nextcloud on and post the answer?

**Comment by e-alfred** [comment]
Okay, I tried it from my server with curl and I am getting the following output:

```
curl "https://nominatim.openstreetmap.org/search.php?q=;;Karl-Wiechert-Allee 10;Hannover;;30604;Deutschland&format=json"

[{"place_id":80086622,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":23444535,"boundingbox":["52.370706","52.3740506","9.805082","9.8078612"],"lat":"52.3722614","lon":"9.8068375","display_name":"Karl-Wiechert-Allee, Kleefeld, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"secondary","importance":0.6199999999999999},{"place_id":75082657,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":13970258,"boundingbox":["52.3747358","52.3785426","9.8080838","9.8091165"],"lat":"52.3765835","lon":"9.8086163","display_name":"Karl-Wiechert-Allee, Heideviertel, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"secondary","importance":0.6199999999999999},{"place_id":254213023,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":672382805,"boundingbox":["52.3921379","52.3921427","9.7982613","9.7984202"],"lat":"52.3921379","lon":"9.7984202","display_name":"Karl-Wiechert-Allee, Groß-Buchholz, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"secondary","importance":0.6199999999999999},{"place_id":69461819,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":4397982,"boundingbox":["52.3922216","52.3936887","9.7931372","9.7968434"],"lat":"52.3928673","lon":"9.7948101","display_name":"Karl-Wiechert-Allee, Groß-Buchholz, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30655, Deutschland","class":"highway","type":"secondary","importance":0.6199999999999999},{"place_id":148923203,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":315072865,"boundingbox":["52.3861937","52.3922052","9.8002465","9.8086979"],"lat":"52.3905189","lon":"9.8056017","display_name":"Karl-Wiechert-Allee, Hannover Rück, Groß-Buchholz, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"secondary","importance":0.6199999999999999},{"place_id":148406462,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":315073734,"boundingbox":["52.3922052","52.392258","9.7984263","9.8002465"],"lat":"52.3922284","lon":"9.7996297","display_name":"Karl-Wiechert-Allee, Lagerplatz Tiefbauamt, Groß-Buchholz, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"secondary","importance":0.6199999999999999},{"place_id":100799209,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":112732601,"boundingbox":["52.3827162","52.3830824","9.8104943","9.8110156"],"lat":"52.3828077","lon":"9.8107073","display_name":"Karl-Wiechert-Allee, Heideviertel, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"tertiary","importance":0.6199999999999999},{"place_id":83026783,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":30140171,"boundingbox":["52.3843397","52.3843956","9.8096836","9.8098932"],"lat":"52.3843956","lon":"9.8098932","display_name":"Karl-Wiechert-Allee, Groß-Buchholz, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"tertiary","importance":0.6199999999999999},{"place_id":149839605,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":315073737,"boundingbox":["52.3922616","52.3926533","9.7982695","9.798275"],"lat":"52.3924344","lon":"9.7982726","display_name":"Karl-Wiechert-Allee, Groß-Buchholz, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"residential","importance":0.6199999999999999},{"place_id":148406428,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"way","osm_id":315073482,"boundingbox":["52.392258","52.392772","9.7983017","9.7984263"],"lat":"52.392614","lon":"9.7983667","display_name":"Karl-Wiechert-Allee, Lagerplatz Tiefbauamt, Groß-Buchholz, Buchholz-Kleefeld, Hannover, Region Hannover, Niedersachsen, 30625, Deutschland","class":"highway","type":"residential","importance":0.6199999999999999}]
```

**Comment by tacruc** [comment]
Wired, 
```php
private function lookupAddressExternal($adr) {
        if (time() - intval($this->config->getAppValue('maps', 'lastAddressLookup')) >= 1) {
            $opts = [
                'http' => [
                    'method' => 'GET',
                    'user_agent' => 'Nextcloud Maps app',
                ]
            ];
            $context = stream_context_create($opts);
            $result_json = @file_get_contents(
                'https://nominatim.openstreetmap.org/search.php?q='.urlencode($adr).'&format=json',
                false,
                $context
            );
            if ($result_json !== false) {
                $result = \json_decode($result_json, true);
                if (!(key_exists('request_failed', $result) AND $result['request_failed'])) {
                    $this->logger->debug('External looked up address: ' . $adr . ' with result' . print_r($result, true));
                    $this->config->setAppValue('maps', 'lastAddressLookup', time());
                    if (sizeof($result) > 0) {
                        if (key_exists('lat', $result[0]) AND
                            key_exists('lon', $result[0])
                        ) {
                            return [
                                $result[0]['lat'],
                                $result[0]['lon'], true
                            ];
                        }
                    }
                    return [null, null, true];
                }
            }
            $this->logger->debug('Externally looked failed');
        }
        return [null, null, False];
```

Basicly I assume that this function is called as you have the "Externally looked failed" in your log. The reply of the nominatim seams as least to be fine from that server. But then there is not so much more which can go wrong.

**Comment by e-alfred** [comment]
Could the missing contacts be related to this bug: https://github.com/nextcloud/contacts/issues/1279
